File: src/SocketManager.vb
Imports System.Net
Imports System.Net.Sockets
Imports System.Threading
Imports Naomai.UTT.Indexer.JulkinNet

''' <summary>
''' UDP socket abstraction and host filter
''' </summary>
Public Class SocketManager
    Protected socket As Socket
    Protected ignoreIps As New List(Of String)

    Protected ipPacketQueue As New Dictionary(Of IPEndPoint, EndpointPacketBuffer)
    Protected lastProcessed As Date

    Public updateIntervalMs As Integer = 100
    Public packetReceiveThread As Thread

    Public Event NewDataReceived(packetBuffer As EndpointPacketBuffer, source As IPEndPoint)

    Public Sub New()
        Dim bindEndpoint As New IPEndPoint(IPAddress.Any, 0)
        socket = New Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp)
        socket.Bind(bindEndpoint)
        socket.ReceiveTimeout = 5000
        socket.ReceiveBufferSize = 1024 * 1024

        lastProcessed = Date.UtcNow

        ReceiveLoop()
        DispatchLoop()
    End Sub

    Public Sub SendTo(endpoint As String, packet As String)
        Dim buffer = System.Text.Encoding.ASCII.GetBytes(packet)

        Dim host As String = GetHost(endpoint), port As UInt16 = GetPort(endpoint)

        Dim addr As IPAddress = Nothing, endpointObj As IPEndPoint = Nothing
        If Net.IPAddress.TryParse(host, addr) Then
            endpointObj = New IPEndPoint(addr, port)
        Else
            endpointObj = New IPEndPoint(Dns.GetHostEntry(host).AddressList(0), port)
        End If


        Try
            socket.SendTo(buffer, endpointObj)
        Catch E As Exception
            Try
                socket.Receive(Nothing, socket.Available, SocketFlags.None)
            Catch NullEx As ArgumentNullException

            End Try
        End Try
    End Sub

    Protected Async Sub ReceiveLoop()
        Dim buffer() As Byte, bindEndpoint As EndPoint = New IPEndPoint(IPAddress.Any, 0)
        Dim sourceEndpoint As IPEndPoint, sourceIp As String
        Dim receiveResult As SocketReceiveFromResult
        Do
            ReDim buffer(2000)
            receiveResult = Await socket.ReceiveFromAsync(buffer, bindEndpoint)

            sourceEndpoint = receiveResult.RemoteEndPoint
            sourceIp = sourceEndpoint.ToString
            If ignoreIps.Contains(sourceIp) Then
                Continue Do
            End If

            ReDim Preserve buffer(receiveResult.ReceivedBytes)
            EnqueuePacket(sourceEndpoint, buffer)


            'Tick()
        Loop
    End Sub

    Protected Async Sub DispatchLoop()
        Do
            Tick()
            Await Task.Delay(25)
        Loop
    End Sub

    Public Sub Tick()
        If (Date.UtcNow - lastProcessed).TotalMilliseconds > updateIntervalMs Then
            DequeueAll()
            lastProcessed = Date.UtcNow
        End If
    End Sub

    Private Sub EnqueuePacket(sourceEndpoint As IPEndPoint, packet As Byte())
        SyncLock ipPacketQueue
            If Not ipPacketQueue.ContainsKey(sourceEndpoint) Then
                ipPacketQueue.Add(sourceEndpoint, New EndpointPacketBuffer)
            End If
        End SyncLock
        ipPacketQueue(sourceEndpoint).Enqueue(packet)
    End Sub

    Private Sub DequeueAll()
        Dim hosts As List(Of IPEndPoint)
        SyncLock ipPacketQueue
            If ipPacketQueue.Count = 0 Then
                Return
            End If

            hosts = ipPacketQueue.Keys.ToList()
        End SyncLock
        For Each host In hosts
            DequeueForHost(host)
        Next
    End Sub

    Private Sub DequeueForHost(host As IPEndPoint)
        Dim packetQueue As EndpointPacketBuffer = ipPacketQueue(host)
        If packetQueue.newPackets Then
            RaiseEvent NewDataReceived(packetQueue, host)
        End If
    End Sub

    Public Sub AddIgnoredIp(ip As EndPoint)
        AddIgnoredIp(ip.ToString)
    End Sub

    Public Sub AddIgnoredIp(ip As String)
        ignoreIps.Add(ip)
    End Sub

    Public Sub ClearIgnoredIps()
        ignoreIps.Clear()
    End Sub

End Class


Public Class EndpointPacketBuffer
    Private packetQueue As New Queue(Of Byte())
    Public newPackets As Boolean = False

    Public Sub Enqueue(packet As Byte())
        SyncLock packetQueue
            packetQueue.Enqueue(packet)
        End SyncLock
        newPackets = True
    End Sub

    Public Function Dequeue() As Byte()
        newPackets = False
        SyncLock packetQueue
            Return packetQueue.Dequeue()
        End SyncLock
    End Function

    Public ReadOnly Property Length As Integer
        Get
            Dim bytesTotal As Integer = 0
            SyncLock packetQueue
                For Each packet In packetQueue
                    bytesTotal += packet.Length
                Next
            End SyncLock
            Return bytesTotal
        End Get
    End Property

    Public Function PeekLast() As Byte()
        newPackets = False
        Return packetQueue.Last()
    End Function

    Public Function PeekAll() As Byte()
        'establish working length of packet stream
        Dim result(Me.Length) As Byte
        Dim offset As Integer = 0
        newPackets = False
        SyncLock packetQueue
            For Each packet In packetQueue
                ' if packet is zero-terminated, grab position of null byte and trim packet
                Dim zeroByteOffset = Array.IndexOf(packet, 0)
                If zeroByteOffset <> -1 Then
                    ReDim Preserve packet(zeroByteOffset) ' also trims null byte
                End If
                packet.CopyTo(result, offset)
                offset += packet.Length
            Next
        End SyncLock
        'adjust length to discard trimmed bytes in zero-terminated packets 
        ReDim Preserve result(offset)
        Return result
    End Function

    Public Sub Clear()
        newPackets = False
        SyncLock packetQueue
            packetQueue.Clear()
        End SyncLock
    End Sub

    Public Overrides Function ToString() As String
        Return "PB:CT=" + packetQueue.Count
    End Function
End Class

File: src/IndexerSharp/SocketManager.cs

using System.Net;
using System.Net.Sockets;
using System.Text;

namespace Naomai.UTT.Indexer;

public class SocketManager <T> where T : ISocket, new()
{
    public readonly T socket;
    protected List<string> ignoreIps = new List<string>();

    protected Dictionary<IPEndPoint, EndpointPacketBuffer> ipPacketQueue = new Dictionary<IPEndPoint, EndpointPacketBuffer>();
    protected DateTime lastProcessed;
    readonly object _queueLock = new object();

    public int UpdateIntervalMs = 100;
    public delegate void NewDataReceivedHandler( EndpointPacketBuffer packetBuffer, IPEndPoint source);
    public event NewDataReceivedHandler? NewDataReceived;

    public SocketManager()
    {
        IPEndPoint bindEndpoint = new IPEndPoint(IPAddress.Any, 0);
        socket = new T();
        
        socket.Bind(bindEndpoint);
        socket.ReceiveTimeout = 5000;
        socket.ReceiveBufferSize = 1024 * 1024;

        lastProcessed = DateTime.UtcNow;

        ReceiveLoop();
        DispatchLoop();
    }

    protected async void ReceiveLoop()
    {
        byte[] buffer = new byte[2000];
        EndPoint bindEndpoint = new IPEndPoint(IPAddress.Any, 0);
        IPEndPoint sourceEndpoint;
        string sourceIp;
        SocketReceiveFromResult receiveResult;

        while(true)
        {
            receiveResult = await socket.ReceiveFromAsync(buffer, bindEndpoint);

            sourceEndpoint = (IPEndPoint)receiveResult.RemoteEndPoint;
            sourceIp = sourceEndpoint.ToString();

            if (ignoreIps.Contains(sourceIp))
            {
                continue;
            }
            EnqueuePacket(sourceEndpoint, 
                    buffer.Take(receiveResult.ReceivedBytes).ToArray()
                );
        }
    }

    public void SendTo(string destination, string packet)
    {
        byte[] buffer = Encoding.ASCII.GetBytes(packet);
        string host = GetHost(destination);
        UInt16 port = GetPort(destination);

        IPAddress addr;
        IPEndPoint endpoint;

        if(!IPAddress.TryParse(host, out addr))
        {
            addr = Dns.GetHostEntry(host).AddressList[0];
        }
        endpoint = new IPEndPoint(addr, port);

        try
        {
            socket.SendTo(buffer, endpoint);
        }
        catch(Exception e)
        {
            FlushSocket();
        }
    }

    protected async void DispatchLoop()
    {
        while (true)
        {
            Tick();
            await Task.Delay(25);
        }
    }

    protected void Tick()
    {
        if ((DateTime.UtcNow - lastProcessed).TotalMilliseconds > UpdateIntervalMs)
        {
            DequeueAll();
            lastProcessed = DateTime.UtcNow;
        }
    }

    private void EnqueuePacket(IPEndPoint sourceEndpoint, byte[] packet)
    {
        lock(_queueLock)
        {
            if(!ipPacketQueue.ContainsKey(sourceEndpoint))
            {
                ipPacketQueue[sourceEndpoint] = new EndpointPacketBuffer();
            }
        }
        ipPacketQueue[sourceEndpoint].Enqueue(packet);
    }

    private void DequeueAll()
    {
        List<IPEndPoint> hosts;

        lock (_queueLock)
        {
            if(ipPacketQueue.Count == 0)
            {
                return;
            }
            hosts = ipPacketQueue.Keys.ToList();
        }
        foreach(IPEndPoint host in hosts)
        {
            DequeueForHost(host);
        }
    }

    private void DequeueForHost(IPEndPoint host)
    {
        EndpointPacketBuffer packetQueue = ipPacketQueue[host];
        if (packetQueue.NewPackets)
        {
            NewDataReceived(packetQueue, host);
        }
    }

    public void AddIgnoredIp(EndPoint ip)
    {
        AddIgnoredIp(ip.ToString());
    }

    public void AddIgnoredIp(string ip)
    {
        ignoreIps.Add(ip);
    }

    public void ClearIgnoredIps()
    {
        ignoreIps.Clear();
    }

    private static string GetHost(string addr)
    {
        string[] slices = addr.Split(":", 2);
        return slices[0];
    }

    private static UInt16 GetPort(string addr)
    {
        string[] slices = addr.Split(":", 2);
        return UInt16.Parse(slices[1]);
    }

    private void FlushSocket()
    {
        try
        {
            socket.Receive(null, socket.Available, SocketFlags.None);
        }
        catch (ArgumentNullException nullEx)
        {

        }
    }
}

public class SocketManager: SocketManager<UdpSocketAdapter>
{

}

public class UdpSocketAdapter :  ISocket
{
    Socket _socket;
    public UdpSocketAdapter()
    {
        _socket = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp);
    }


    public int ReceiveBufferSize
    {
        get => _socket.ReceiveBufferSize;
        set => _socket.ReceiveBufferSize = value;
    }

    public int ReceiveTimeout
    {
        get => _socket.ReceiveTimeout;
        set => _socket.ReceiveTimeout = value;
    }

    public int Available => _socket.Available;

    public virtual void Bind(EndPoint localEP) => _socket.Bind(localEP);

    public virtual int Receive(byte[] buffer, int size, SocketFlags socketFlags)
        => _socket.Receive(buffer, size, socketFlags);

    public virtual async Task<SocketReceiveFromResult> ReceiveFromAsync(ArraySegment<byte> buffer, EndPoint remoteEP)
        => await _socket.ReceiveFromAsync(buffer,  remoteEP);
    

    public virtual int SendTo(byte[] buffer, EndPoint remoteEP) => _socket.SendTo(buffer, remoteEP);
}

public interface ISocket
{
    int ReceiveTimeout { get; set; }
    int ReceiveBufferSize { get; set; }
    int Available { get; }

    void Bind(EndPoint localEP);
    int Receive(byte[] buffer, int size, SocketFlags socketFlags);
    Task<SocketReceiveFromResult> ReceiveFromAsync(ArraySegment<byte> buffer, EndPoint remoteEP);
    int SendTo(byte[] buffer, EndPoint remoteEP);
}


public class EndpointPacketBuffer
{
    Queue<byte[]> _packetQueue = new Queue<byte[]>();
    readonly object _queueLock = new object();
    public bool NewPackets = false;

    public void Enqueue(byte[] packet)
    {
        lock (_queueLock){
            _packetQueue.Enqueue(packet);
        }
        NewPackets = true;
    }

    public byte[] Dequeue()
    {
        NewPackets = false;
        lock (_queueLock)
        {
            return _packetQueue.Dequeue();
        }
    }

    public byte[] PeekLast()
    {
        NewPackets = false;
        return _packetQueue.Last();
    }

    public byte[] PeekAll()
    {
        // establish working length of packet stream
        byte[] result = new byte[this.Length];
        int offset = 0;
        NewPackets = false;

        lock (_queueLock)
        {
            foreach (byte[] packet in _packetQueue)
            {
                // if packet is zero-terminated, grab position of null byte and trim packet
                byte[] packetTrimmed = packet;
                int zeroByteOffset = Array.IndexOf(packet, 0);
                if(zeroByteOffset != -1)
                {
                    Array.Resize(ref packetTrimmed, zeroByteOffset);
                }
                packetTrimmed.CopyTo(result, offset);
                offset += packetTrimmed.Length;
                
            }
        }
        // adjust length to discard trimmed bytes in zero-terminated packets 
        Array.Resize(ref result, offset);
        return result;
    }

    public void Clear()
    {
        NewPackets = false;
        lock (_queueLock)
        {
            _packetQueue.Clear();
        }
    }

    public int Length
    {
        get
        {
            int bytesTotal = 0;
            lock (_queueLock)
            {
                foreach (byte[] packet in _packetQueue)
                {
                    bytesTotal += packet.Length;
                }
            }
            return bytesTotal;
        }
    }

    public override string ToString()
    {
        return $"PB:CT={_packetQueue.Count}";
    }
}
